File: CPUEmulation/MinimalRAM.swift
//
//  MinimalRAM.swift
//  CPUEmulation
//
//  Created by Will Stafford on 8/16/17.
//  Copyright © 2017 Wrsford. All rights reserved.
//

import Foundation

public class MinimalRAM: EmuMemory {
    
    enum MinimalRamError: Error {
        case noMemoryRegion(virtualAddress: EmuInt)
    }
    
    class Region {
        var virtualAddress: EmuInt
        static let size: EmuInt = 2048
        let size: EmuInt = Region.size
        private var backingBytes: [EmuByte] = [EmuByte](repeating: 0, count: Region.size)
        
        init(virtual: EmuInt) {
            // we are going to align the virtual address here
            let alignedAddress: EmuInt = virtual - (virtual % size)
            virtualAddress = alignedAddress
        }
        
        func adjustedAddress(_ address: EmuInt) -> EmuInt
        {
            return address - virtualAddress
        }
        
        subscript(address: EmuInt) -> EmuByte {
            get {
                return backingBytes[adjustedAddress(address)]
            }
            set(newValue) {
                backingBytes[adjustedAddress(address)] = newValue
            }
        }
    }
    
    private var regions: [Region] = []
    
    /// Gets the region that owns the given address
    private func getRegion(address virtualAddress: EmuInt) throws -> Region  {
        var matchingRegion: Region? = nil
        // Check all regions for the address owner
        for region in regions {
            // Check if it is within the region's lower bound
            if (virtualAddress >= region.virtualAddress)
            {
                // Check if it is within the region's upper bound
                if (virtualAddress < region.virtualAddress + region.size)
                {
                    // Found owning region
                    matchingRegion = region
                    break
                }
            }
        }
        
        // No region found
        if (matchingRegion == nil)
        {
            throw MinimalRamError.noMemoryRegion(virtualAddress: virtualAddress)
        }
        else {
            return matchingRegion!
        }
    }
    
    public func setByte(_ val: EmuByte, at address: EmuInt) {
        do {
            let region = try getRegion(address: address)
            // Region exists, set the value
            region[address] = val
            
        } catch MinimalRamError.noMemoryRegion(_) {
            // Need a new region
            
            let newRegion = Region(virtual: address)
            newRegion[address] = val
            regions.append(newRegion)
        }
        catch {
            print("Couldn't set byte at \(address)!")
        }

    }
    
    public func getByte(at address: EmuInt) -> EmuByte {
        do {
            let region = try getRegion(address: address)
            
            return region[address]
            
        } catch MinimalRamError.noMemoryRegion(let virtualAddress) {
            print("No owning region found for requested address: \(virtualAddress)")
            return 0
        }
        catch {
            return 0
        }
    }
}

File: CPUEmulation/minimal_ram.c
//
//  minimal_ram.c
//  CPUEmulation
//
//  Created by Will Stafford on 8/24/17.
//  Copyright © 2017 Wrsford. All rights reserved.
//

#include "minimal_ram.h"

struct minimal_ram_context {
    min_ram_region regions[MIN_RAM_MAX_REGION_COUNT];
    long region_count;
};

// internal defs
extern min_ram_region* get_region_for_address(minimal_ram_context* ram, long address);
extern long has_region_for_address(minimal_ram_context* ram, long address);
extern void add_region_for_address(minimal_ram_context* ram, long address);
extern long get_region_byte_for_virtual_address(min_ram_region* region, long virtual_address);
extern void set_region_byte_for_virtual_address(min_ram_region* region, long virtual_address, long value);

// Public

/// Constructor
void minimal_ram_init(minimal_ram_context* ram)
{
    ram->region_count = 0;
}

/// Dealloc
void minimal_ram_teardown(minimal_ram_context* ram)
{
    
}


/// Gets a byte at the address
long min_ram_get_byte(minimal_ram_context* ram, long address)
{
    min_ram_region* region = get_region_for_address(ram, address);
    
    if (region == NULL)
    {
        // Region doesn't exist
        // I believe the Swift version returns 0 in this case
        return 0;
    }
    else {
        return get_region_byte_for_virtual_address(region, address);
    }
    
}


/// Sets a byte at the address
void min_ram_set_byte(minimal_ram_context* ram, long address, long val)
{
    // Check if the region exists
    if (has_region_for_address(ram, address) == 0)
    {
        // Doesn't exist, add it
        add_region_for_address(ram, address);
    }
    
    // Get the region
    min_ram_region* region = get_region_for_address(ram, address);
    
    // Set the value
    set_region_byte_for_virtual_address(region, address, val);
}


// Private/internal

/// Gets the owning region for the virtual address
min_ram_region* get_region_for_address(minimal_ram_context* ram, long address)
{
    min_ram_region* region = NULL;
    long i;
    for (i = 0; i < ram->region_count; i++)
    {
        // Check the lower boundary
        if (ram->regions[i].virtual_address <= address)
        {
            // Check upper boundary
            if (ram->regions[i].virtual_address + ram->regions[i].size > address)
            {
                // Found a match
                region = &ram->regions[i];
                // Break out of loop & return
                break;
            }
        }
    }
    
    // NULL if it doesn't exist
    return region;
}

/// Checks if a region exists
long has_region_for_address(minimal_ram_context* ram, long address) {
    min_ram_region* region = get_region_for_address(ram, address);
    if (region == NULL)
    {
        // Doesn't exist
        return 0;
    }
    else {
        // It exists
        return 1;
    }
}

/// Adds a region for the address (auto-aligned)
void add_region_for_address(minimal_ram_context* ram, long address)
{
    min_ram_region newRegion;
    
    // Set the size
    newRegion.size = MIN_RAM_REGION_SIZE;
    
    // Requested address offset from an aligned address
    long virtualBaseOffset = address % newRegion.size;
    // The aligned base address
    long alignedBaseAddress = (address - virtualBaseOffset);
    
    // check calculated address
    if (alignedBaseAddress % newRegion.size != 0)
    {
        // Bad calculation
        printf("Bad alignment calculation!");
        // Crash the program so we know something is wrong (tries to get ram address -1)
        newRegion.virtual_address = ((int*)NULL)[-1];
    }
    
    // Set the aligned address
    newRegion.virtual_address = alignedBaseAddress;
    
    // Add it
    ram->regions[ram->region_count++] = newRegion;
}

/// Calculates the actual address for the region and returns the value
long get_region_byte_for_virtual_address(min_ram_region* region, long virtual_address)
{
    // Calc relative address
    long regionAddr = virtual_address - region->virtual_address;
    // Return byte
    return region->values[regionAddr];
}

/// Calculates the actual address for the region and sets the value
void set_region_byte_for_virtual_address(min_ram_region* region, long virtual_address, long value)
{
    // Calc relative address
    long regionAddr = virtual_address - region->virtual_address;
    // Set byte
    region->values[regionAddr] = value;
}
